### PR TITLE
tabledesc: fix incomplete post-deserialization changes

### PR DESCRIPTION
**Pull request by postamar**
After being deserialized from their protobuf representations, table
descriptors, like all descriptors, go through a round of
post-deserialization changes, in order to bring them up to date with
respect to the current codebase.

A canonical example of why this is useful is restoring a backup with
a table using the deprecated foreign-key representation. Instead of
having to support deprecated foreign keys throughout the codebase, we
migrate these to the current representation in this post-deserialization
step.

Another class of post-deserialization changes pertain to index
descriptors and their versions. A bug in how these changes were applied
meant that the changes could sometimes be skipped. This commit fixes
this.

The impact of this bug is limited by the fact that index descriptor
versions are backward-compatible.

Release note: None

### CODE REVIEW

File: pkg/sql/catalog/tabledesc/table_desc_builder.go [hunk 1]
```diff
@@ -192,11 +192,13 @@ func maybeFillInDescriptor(
 
 	changes.UpgradedIndexFormatVersion = maybeUpgradePrimaryIndexFormatVersion(desc)
 	for i := range desc.Indexes {
-		changes.UpgradedIndexFormatVersion = changes.UpgradedIndexFormatVersion || maybeUpgradeSecondaryIndexFormatVersion(&desc.Indexes[i])
```
Comment (postamar) [inline pkg/sql/catalog/tabledesc/table_desc_builder.go:195]: Assuming a table descriptor has several secondary indexes needing an upgrade, short-circuit evaluation would ensure that only the first secondary index would get upgraded 😬 

I wrote this. I'm quite embarrassed.

Comment (ajwerner) [reply pkg/sql/catalog/tabledesc/table_desc_builder.go:195]: Oh man. Glad you caught this. Does this have any implications for needing to retain migrations we might have stripped?

**Comment by postamar** [comment]
I've added another commit, with the view of backporting this PR to 21.1 and 20.2 and so fixing https://github.com/cockroachdb/cockroach/issues/71552

File: pkg/sql/catalog/tabledesc/table_desc_builder.go (hunk 1 above)
Comment (postamar) [reply pkg/sql/catalog/tabledesc/table_desc_builder.go:195]: I don't think so and I wouldn't worry about it. If we need to worry about it, then let's think about putting in a "standing migration" of sorts which round-trips each descriptor in `system.descriptor` through `RunPostDeserializationChanges` at the end of each major version upgrade.

File: pkg/sql/catalog/tabledesc/structured_test.go [hunk 2]
```diff
@@ -408,28 +450,137 @@ func TestMaybeUpgradeIndexFormatVersion(t *testing.T) {
 					},
 				},
 			},
-			expUpgrade:  false,
 			expValidErr: "relation \"tbl\" (51): index \"other\" has column ID 1 present in: [KeyColumnIDs KeySuffixColumnIDs]",
 		},
+		{ // 4
```
Comment (ajwerner) [inline pkg/sql/catalog/tabledesc/structured_test.go:455]: nit: describe what this is exercising. A future reader will be hard-pressed to figure it out from just the descriptor without any commentary

File: pkg/sql/catalog/tabledesc/structured_test.go [hunk 3]
```diff
@@ -408,28 +450,137 @@ func TestMaybeUpgradeIndexFormatVersion(t *testing.T) {
...
+				return
+			}
+
+			if e, a := test.upgraded, desc.TableDesc(); !reflect.DeepEqual(e, a) {
+				for _, diff := range pretty.Diff(e, a) {
+					t.Error(diff)
+				}
+			}
+
+			// Run post-deserialization changes again, descriptor should not change.
```
Comment (ajwerner) [inline pkg/sql/catalog/tabledesc/structured_test.go:574]: nice

**Review by ajwerner — APPROVED** [review]
LGTM

File: pkg/sql/catalog/tabledesc/structured_test.go (hunk 2 above)
Comment (postamar) [reply pkg/sql/catalog/tabledesc/structured_test.go:455]: You're right. I'll add some descriptions.

**Comment by postamar** [comment]
Thanks for the review!

bors r+

File: pkg/sql/catalog/tabledesc/structured_test.go (hunk 3 above)
Comment (postamar) [reply pkg/sql/catalog/tabledesc/structured_test.go:583]: Thanks but I wish I'd tested this stuff properly the first time around and saved us a lot of heartache.

File: pkg/sql/catalog/tabledesc/structured_test.go (hunk 2 above)
Comment (postamar) [reply pkg/sql/catalog/tabledesc/structured_test.go:461]: Done.

**Pull request merged**